### DISCUSSION TITLE
Error "Possibly non-uniform value passed here" after Chrome update

### DISCUSSION THREAD

**Discussion by jay19240** [gpuweb/gpuweb#4332] [Q&A] [answered]
Hello, just after updated Chrome i was surprised to get these errors:
```
Tint WGSL reader failure: :206:18 error: 'textureSample' must only be called from uniform control flow
  var texColor = textureSample(DECAL_ATLAS_TEXTURE, DECAL_ATLAS_SAMPLER, vec2<f32>(uvx, uvy));
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

:189:24 note: control flow depends on possibly non-uniform value
  if (clipPos.z < -1.0 || clipPos.z > 1.0)
                       ^^

:188:53 note: parameter 'fragPos' of 'CalcDecal' may be non-uniform
  var clipPos = DECALS[index].VP_MATRIX * vec4<f32>(fragPos, 1.0);
                                                    ^^^^^^^

:120:37 note: possibly non-uniform value passed here
        decalsColor += CalcDecal(i, FragPos);
                                    ^^^^^^^

:120:37 note: user-defined input 'FragPos' of 'main' may be non-uniform
        decalsColor += CalcDecal(i, FragPos);
```
FragPos is not an uniform, it is a varying but we are able to used varying inside the program so i'm a little bit confused...
To add context, the line 206, 189 and 188 are inside CalcDecal function that is called in my main fragment shader function.
Do you known what is the problem and if new restrictions are added recently to WGSL ?

**Comment by ben-clayton** [comment]
No new restrictions were added, but a bug was recently fixed where UA violations could sometimes not get reported. IIRC correctly this would only happen for shaders where the number of module-scope declarations exceeded a constant.

**Reply by jay19240** [reply]
So what is the concret solution to solve this ?

**Comment by jay19240** [comment]
The following code that cause this error:
```
fn CalcDecal(index: u32, fragPos: vec3<f32>) -> vec4<f32>
{
  var clipPos = DECALS[index].VP_MATRIX * vec4<f32>(fragPos, 1.0);
  if (clipPos.z < -1.0 || clipPos.z > 1.0)
  {
    return vec4(0.0, 0.0, 0.0, 0.0);
  }

  var ndcPos = vec3<f32>(clipPos.xyz / clipPos.w).xy * DECALS[index].ASPECT_RATIO;
  if (ndcPos.x < -1.0 || ndcPos.x > 1.0 || ndcPos.y < -1.0 || ndcPos.y > 1.0)
  {
    return vec4(0.0, 0.0, 0.0, 0.0);
  }

  var uvx = ndcPos.x * 0.5 + 0.5;
  uvx = (uvx * DECALS[index].TEXTURE_WIDTH) + DECALS[index].TEXTURE_LEFT;

  var uvy = 1.0 - (ndcPos.y * 0.5 + 0.5);
  uvy = (uvy * DECALS[index].TEXTURE_HEIGHT) + DECALS[index].TEXTURE_TOP;

  var texColor = textureSample(DECAL_ATLAS_TEXTURE, DECAL_ATLAS_SAMPLER, vec2<f32>(uvx, uvy));
  texColor.a = max(texColor.a - (1.0 - DECALS[index].OPACITY), 0.0);
  return texColor;
}
```
For me it seem good...
If i comment 3 last lines, the error is removed and program works.

**Reply by ben-clayton** [reply]
Please file an issue at [crbug.com/tint](https://crbug.com/tint) with the full WGSL shader code, and we'll take a look and advise.

**Comment by bbbbx** [comment, marked as answer]
Meet the same issue after updating to Chrome 118.

According to the newest spec., `textureSample` function:

> If [uniformity analysis](https://gpuweb.github.io/gpuweb/wgsl/#uniformity-analysis) cannot prove a call to this function is in [uniform control flow](https://gpuweb.github.io/gpuweb/wgsl/#uniform-control-flow), then a [derivative_uniformity](https://gpuweb.github.io/gpuweb/wgsl/#trigger-derivative_uniformity) [diagnostic](https://gpuweb.github.io/gpuweb/wgsl/#diagnostic) is [triggered](https://gpuweb.github.io/gpuweb/wgsl/#triggered).

so, `textureSample` must only be called from uniform control flow, otherwise a derivative_uniformity diagnostic will be triggered.

A [global diagnostic filter](https://gpuweb.github.io/gpuweb/wgsl/#global-diagnostic-filter) can be used to apply a diagnostic filter to the entire WGSL module.

You can use a global diagnostic filter to avoid this behavior, put following line at WGSL shader string top:
```wgsl
diagnostic(off, derivative_uniformity);
```

but I can't confirm if this diagnostic filter hurts performance.

**Reply by jay19240** [reply]
Yes it is the expected behavior as ben said. For my own specific case is solved by removed two first returns and adding a multiplication in the last return with a color variable equal to 0 or 1 depending conditions.
